server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -273,9 +273,15 @@ private Interval findIntervalForKill(
   )
   {
     final DateTime minStartTime = datasourceToLastKillIntervalEnd.get(dataSource);
+
+    // Once the first segment from a datasource is killed, we have a valid minStartTime.
+    // Restricting the upper bound to scan segments metadata while running the kill task results in a efficient SQL query.
     final DateTime maxEndTime = ignoreDurationToRetain
                                 ? DateTimes.COMPARE_DATE_AS_STRING_MAX
-                                : DateTimes.nowUtc().minus(durationToRetain);
+                                : DateTimes.min(DateTimes.nowUtc().minus(durationToRetain),
+                                minStartTime != null
+                                        ? minStartTime.plus(maxIntervalToKill)
+                                        : DateTimes.nowUtc().minus(durationToRetain));
```

Comment:
This is a little difficult to read. Please assign this value to a variable beforehand.